design/design.md

```diff
@@ -0,0 +1,414 @@
+a multisignature scheme, where the data elements would specifically be
+signatures.
+Thus there is question of whether we want our library to provide a general
+ALBA interface or simply an interface for a multisignature scheme.
+We will do both.
+1. We will implement the most generic version of ALBA to make it suitable for
+as many applications as possible.
+This means that there is only a generic notion of "data elements" and no notion
+of signatures, VRF, etc.
+2. We will use the generic ALBA library to implement a multisignature scheme,
+either as a small example in this repository or a full-fledged library in a
+separate repository.
+Either way, the multi-signature implementation would serve as a flagship
+example of how the ALBA library can be used.
+
+The requirements for our (general) ALBA implementation can be roughly split
+in four groups: 
+1. definitions: what do the ALBA schemes we implement guarantee, under what
+assumptions?
+2. correctness: does our implementation provide the same guarantees of
+soundness and completeness as the ALBA schemes on paper?
+3. performance: proof size, communication complexity, computational complexity.
+4. interfaces: the applications should be able to use our
+implementation relatively easily
+
+### Definitions
+
+The definitions requirement can be stated as follows.
+Taking an ALBA protocol we implement, can you construct (on paper) a larger
+protocol, such as Mithril, and mathematically prove its guarantees such as
+completeness and soundness?
+Currently, ALBA schemes satisfying the definitions in the ALBA paper cannot
+be used to construct Mithril, Peras and Leios.
+
+First, all three require a multisignature scheme secure against
+grinding attacks.
+In short, unlike the setting in the ALBA paper, the theorem statement to be
+proven is not fixed in advance, but can depend on the previous actions of
+the adversary.
+The adversary can simulate (grind) different actions to see if there is one
+that will allow him to cheat.
+For instance, the lottery in Mithril uses a seed from the Ouroboros protocol
+to pseudo-randomly select the signing committee;
+this seed can to some extent be controlled by the adversary.
+Research is needed to see how grinding can be handled and on what
+level: ALBA or the multisignature scheme.
+
+Additionally, Mithril and possibly Peras and Leios cannot provide a weight
+oracle as required by the definitions in the ALBA paper.
+The weight oracle must report the weight of any data element.
+For Mithril, if the data elements are signatures, the weight oracle would
+have to report for any signature the amount of stake of the creator of the
+signature.
+Clearly, this is impossible.
+Even if the data element was a pair of public key and signature, the weight
+oracle would have no way of verifying that the public key owns any stake.
+Again, research is needed to construct the right ALBA scheme on
+paper before implementing it.
+
+### Correctness
```

Comment:
There will _always_ be a discrepancy because of types and what's not. There is no "ideal", we will implement the scheme with that goal using best practices, you may note that there is a discrepancy between theory and practice.
You may list where the errors are, but be more concise, and don't use "tiny".

Reply:
Because of precision. I could ask for a security paramater equals to E and no computer would be able to give me the exact result. 

Regardless, our goal is to balance correctness, readability and security of the protocol. This is a reference implementation hence we care about the protocol being correct, not the rounding and other errors.

tiny: I have not seen it in papers, I usually see "not significant", so I have been assuming it to be rather informal